rgw/sfs: Object Lock - metadata changes

giubacc opened the pull request:
I'd like to proceed by relatively small steps.
It could be necessary to make additional adjustments to metadata in subsequent PRs.
**These changes to metadata are NOT retro-compatible.**

Fixes: https://github.com/aquarist-labs/s3gw/issues/312
Signed-off-by: Giuseppe Baccini <giuseppe.baccini@suse.com>


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

giubacc commented on src/test/rgw/sfs/test_rgw_sfs_sqlite_buckets.cc:
```diff
@@ -19,6 +19,39 @@ using namespace rgw::sal::sfs::sqlite;
 namespace fs = std::filesystem;
 const static std::string TEST_DIR = "rgw_sfs_tests";
 
+/*
```
@jecluis @0xavi0 what do you think about this approach?
It is an hack to allow to fill, otherwise not fillable structs.
This is for testing only obviously.

jecluis replied:
I must say I don't have an opinion on this. If you believe this to be the right approach to mock what needs to be mocked, I'm okay with it.

giubacc commented:
@0xavi0 as per chat conversation, there is the test regarding the schema compatibility that is failing.
This is supposed to fail for this PR, because the changes are not retro-compatible.

giubacc commented on src/rgw/driver/sfs/multipart.cc:
```diff
@@ -24,10 +24,14 @@ using namespace std;
+  auto mmo = std::make_unique<SFSMultipartMetaObject>(
+      store, rgw_obj_key(get_meta(), string(), RGW_OBJ_NS_MULTIPART),
+      bucket, bucketref
+    );
```
@0xavi0 @jecluis @irq0 these following three lines should represent the fix for the problem we were debugging.
There is no need to implement complete persistence for attributes in the multipart upload context (for this PR to work).
`get_meta_obj()` needed to be updated to reflect the attributes to the returned `mmo` obj.

jecluis replied:
lgtm, but this should be in a different commit. It's a different logical change than the rest of the patch. I would suggest spinning this code to a new patch and have it before the remaining code changes in the git history.

jecluis commented on src/rgw/driver/sfs/CHANGELOG.md:
```diff
@@ -7,7 +7,23 @@ The format is based on
 and this project adheres to
 [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [0.11.0] - Unreleased
```
let's drop this change, and add the release notes to the wiki at https://github.com/aquarist-labs/s3gw/wiki/Current-CHANGELOG

jecluis commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -23,15 +23,15 @@ using namespace std;
 ) {
-
-  source->refresh_meta();
-  objref = source->get_object_ref();
```
what's the reasoning behind moving this to the constructor?

giubacc replied:
I cannot remember the exact case, but keeping this into `prepare` is not sufficient in all cases.
There are pieces of SAL code that are calling `get_*()` methods but they don't call `prepare()`.
In these cases the  `SFSReadOp` is not properly initialized and those calls are going to fail.
As solution to this I moved that initializing code into the constructor to cover all the cases.

jecluis replied:
ack. Mind adding a comment on the constructor saying that?

jecluis commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -282,26 +279,91 @@ void SFSObject::gen_rand_obj_instance_name() {
-  rgw_obj *target_obj
-) {
-  lsfs_dout(dpp, 10) << " TODO" << dendl;
+/*
+This must be intended as a fetch of object's attributes.
+Caller then will likely call get_attrs().
+*/
+int SFSObject::get_obj_attrs(optional_yield y,
+                             const DoutPrefixProvider *dpp,
+                             rgw_obj *target_obj) {
+  refresh_meta();
```
Is there no expectation that the object be returned in `target_obj`? It seems weird that we have a `get` operation that is just refreshing metadata.

giubacc replied:
This is how I have interpreted this call.
I have checked around and the usage seems to suggest that one first calls `get_obj_attrs` and then it uses the `get_attrs`.
I have serious doubts that this guess is entirely correct for all cases; so for the time being I would leave `target_obj` empty and rely on the fact that in the future something could potentially fail because of this. But this failing will have a more precise semantics that a potentially fuzzy failing because now we are filling `target_obj` with what we think is our current guess.

jecluis replied:
if that's the case, I'd suggest adding a comment about that in the function, and assert in case your expectations are not met. That way we can potentially catch corner cases.

jecluis commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -54,16 +54,13 @@ int SFSObject::SFSReadOp::get_attr(const DoutPrefixProvider *dpp,
-    policy.encode(dest);
-    return 0;
+  if (!objref || objref->deleted) {
+    return -ENOENT;
   }
-  return -ENOTSUP;
+  if(!objref->get_attr(name, dest)){
+    return -ENODATA;
+  }
```
there's inconsistent formatting throughout the patchset when it comes to spacing. I understand this is a personal preference thing, but if so it would be nice to keep the same throughout the patchset. Even though there are coding guidelines for the Ceph project (which we have been very good at not following) to which we should aim adhering to.

giubacc replied:
I agree with you that we should try to follow the same convention (and I'm trying) and in this case I will fix this (because you point me out).
Unfortunately, this is something I'm not good at all without an automatic tool that is doing this kind of work for me.
I will always do mistakes of this kind (This is why I tend to rely completely on formatting tools).
So expect my code to be potentially inconsistent from time to time.

jecluis replied:
If you find a formatting tool that allows us to just format code that has been edited, that would be awesome. We could define the rules for everybody, and have it go through the checks to ensure code conformity. We just can't be doing that for the entire code base :\

jecluis commented on src/rgw/driver/sfs/sqlite/objects/object_definitions.h:
```diff
@@ -43,8 +41,6 @@ struct DBOPObjectInfo {
   ceph::real_time mtime;
   ceph::real_time set_mtime;
   ceph::real_time delete_at;
-  rgw::sal::Attrs attrs;
```
why are we removing the object attrs from the object?

giubacc replied:
We are moving  `attrs` from `objects` to `versioned_objects`.
This because every version of an object is a logically separate entity from every other version.
Every version has its own attributes.
Non versioned objects have also a row inside `versioned_objects` (it can be intended as the only existing version).
moving `attrs` to `versioned_objects` uniforms the code for both `object` and `versioned-object` types.

jecluis replied:
ack

jecluis commented on src/rgw/driver/sfs/user.cc:
```diff
@@ -174,6 +174,15 @@ int SFSUser::create_bucket(
+    https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-overview.html#object-lock-bucket-config
+  */
+  if (obj_lock_enabled)
+    info.flags = BUCKET_VERSIONED | BUCKET_OBJ_LOCK_ENABLED;
```
shouldn't this be OR'ed with the existing flags? How do we know we're not squashing existing flags?

giubacc replied:
This is a reasonable observation in general and if you think, we can enforce the safety ORing the existing flag.
I have doubts that in this case this could actually happen because:

1) this is a create operation, we are defining the flags inside this function.
2) I've copied this from `rados` implementation.

Commit:
rgw/sfs: Object Lock - metadata changes

- buckets table
   - flags:
     - BUCKET_OBJ_LOCK_ENABLED
     - BUCKET_VERSIONED
     - both encoded into existing buckets.flags field
   -  add 1 BLOB field:
     - object_lock: this holds the default object lock configuration for the bucket.
- objects table
   - drop acls column: ACLs are stored into attrs field, no need to have
     a dedicated field.
   - drop attrs column: non-versioned object will use the versioned_objects.attrs field too.
- versioned_objects table
   - add attrs column
- adequated code for reading/storing object's attributes

Fixes: https://github.com/aquarist-labs/s3gw/issues/312
Signed-off-by: Giuseppe Baccini <giuseppe.baccini@suse.com>

Commit:
rgw/sfs: set the attributes on SFSMultipartUpload::get_meta_obj()

Set the attributes on the object returned by SFSMultipartUpload::get_meta_obj()

Signed-off-by: Giuseppe Baccini <giuseppe.baccini@suse.com>

giubacc commented:
@jecluis are we good ? can we merge this PR?

irq0 commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -23,15 +23,23 @@ using namespace std;
+    In those cases the SFSReadOp is not properly initialized and those
+    calls are going to fail.
+  */
+  source->refresh_meta();
```
Any side effects to this? refresh_meta makes a database access and may throw

giubacc replied:
are you suggesting to `try..catch()` every occurrence of `refresh_meta` in `sfs` code?

irq0 replied:
Not suggesting that. Asking what the side effects are to moving that to the constructor. Possible errors will now occur there not in prepare. Is that a problem?

giubacc replied:
unfortunately hard to say; I have no idea if that would have been a problem even in the previous location.

irq0 commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -282,26 +287,98 @@ void SFSObject::gen_rand_obj_instance_name() {
-  lsfs_dout(dpp, 10) << " TODO" << dendl;
+/*
+  This should be intended as a mere fetch of object's attributes.
+  The caller will likely call get_attrs() after get_obj_attrs().
```
get_attr?

giubacc replied:
can you please elaborate more precisely your question?

irq0 replied:
Is get_attrs a typo? There is no such function in object.cc. There is however a get_attr

giubacc replied:
no typo: there are 2 distinct functions:

- `get_attrs` (SAL)
- `get_attr` (internal)

irq0 commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -282,26 +287,98 @@ void SFSObject::gen_rand_obj_instance_name() {
+  This should be intended as a mere fetch of object's attributes.
+  The caller will likely call get_attrs() after get_obj_attrs().
+  This is how we have interpreted this call.
+  The average usage of this seems to suggest that one first calls
```
what is average usage?

giubacc replied:
Not sure to have understood correctly what are you asking here, btw: for "average usage" is what I have interpreted as "a frequent usage case" for `get_obj_attrs` method.

irq0 replied:
The sentence is worded very vague and passive. We can know how calls what. We should know. Especially if we derive calling semantics from that. So, what is the average usage? There is possibly just one usage in the RGWOps

giubacc replied:
the problem is that the sentence `is worded very vague and passive` because it is! :grin:
It represents perfectly our uncertainty.

giubacc replied:
PS:feel  free to investigate deeper the topic if you think it is worth to.

irq0 commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -282,26 +287,98 @@ void SFSObject::gen_rand_obj_instance_name() {
+  so for the time being, we leave target_obj empty.
+  We rely on the fact that, if in the future something could potentially fail
+  because target_obj is left empty, that fail will be explicit.
+*/
```
Hard to grok. Can we make this more concise? 
My interpretation is:
1. We expect get_attr calls after get_obj_attrs calls. Can we assert this in code?
2. We don't set target_obj. If a caller uses that, boom

Would a dummy / sentinel target_obj make sense?

giubacc replied:
Please, let's always take in consideration that we are not the original designers of the abstract code, therefore we cannot always argue 100% precisely the original semantics as was originally thought by original designers. 
The process of interpretation is by definition uncertain.
Accepting to work with a codebase that was designed by others necessarily leads to compromises; also in terms of interface semantics guessing.
In my opinion an assured boom in an uncertain scenario derived by a lack of knowledge is better than an hiding of that.
Hiding something that is uncertain, IMO is worst that explicitly make it fail.

jecluis replied:
@irq0 I think the requirements to asserting on "get_attrs after get_obj_attrs" would be non-trivial and potentially for little to no gain.

I think a more interesting assert would be on `target_obj`. From the call stack, do we know whether it's expected to be populated when the function is called?

giubacc replied:
Since the signature (RADOS BE) of `get_obj_attrs` is:

```
virtual int get_obj_attrs(optional_yield y, const DoutPrefixProvider* dpp, rgw_obj* target_obj = NULL) override;
```

We should have no doubts that this function can be called with 2 or 3 args.
It's hard to say if there is a rule here, we should check every call to this function to check if, when `target_obj` is not `NULL`, this is an empty object, or instead is a dirty one.

irq0 replied:
> @irq0 I think the requirements to asserting on "get_attrs after get_obj_attrs" would be non-trivial and potentially for little to no gain.

ack

irq0 commented on src/rgw/driver/sfs/object.cc:
```diff
@@ -282,26 +287,98 @@ void SFSObject::gen_rand_obj_instance_name() {
                                        optional_yield y) {
-  ldpp_dout(dpp, 10) << __func__ << ": TODO" << dendl;
-  return -ENOTSUP;
+  if(!attr_name){
```
Why would the upper layer call this with an empty string attr name?

giubacc replied:
I think I have copied this from another BE implementation; 

1.  does it make really sense to check `attr_name`? probably not.
2. it is really important in terms of computational cost? probably not.

irq0 replied:
Is an empty string a special value perhaps? 

If it probably doesn't make sense to do this, I suggest finding that out.

giubacc replied:
I honestly think this is not something worth to be addressed.
As above, feel free to investigate better if you believe.

irq0 commented on src/rgw/driver/sfs/types.h:
```diff
@@ -73,8 +73,12 @@ struct Object {
   int delete_object_version(SFStore *store);
   void delete_object(SFStore *store);
+
+  bool get_attr(std::string name, bufferlist& dest);
```
const std::string&?

giubacc replied:
will change

irq0 commented on src/rgw/driver/sfs/object.h:
```diff
@@ -218,28 +219,30 @@ class SFSObject : public StoreObject {
+                            Attrs *delattrs,
+                            optional_yield y) override;
+
+  bool get_attr(std::string name, bufferlist& dest);
```
const std::string&?

giubacc replied:
will change

jecluis commented:
> @jecluis are we good ? can we merge this PR?

I think it would be better to wait for the release branches to be cut, and merge it for the next iteration (v0.13.0).
